packages/beacon-state-transition/src/cache/stateCache.ts

```diff
@@ -0,0 +1,151 @@
+ * - they first deposit
+ * - they dip from 32 effective balance to 31 (pretty much only when inactive for very long, or slashed)
+ * - they activate (once)
+ * - they exit (once)
+ * - they get slashed (max once)
+ *
+ * ### balances
+ * The balances array completely changes at the epoch boundary, where almost all the validator balances
+ * are updated. However it may have tiny changes during block processing if:
+ * - On a valid deposit
+ * - Validator gets slashed
+ * - On altair, the block proposer. Optimized to only happen once per block
+ *
+ * ### epochParticipation
+ * epochParticipation changes continuously through the epoch for each partipation bit of each valid attestation in the state.
+ * The entire structure is dropped after two epochs.
+ *
+ * ### inactivityScores
+ * inactivityScores can be changed only:
+ * - At the epoch transition. It only changes when a validator is offline. So it may change a bit but not
+ *   a lot on normal network conditions.
+ * - During block processing, when a validator joins a new 0 entry is pushed
+ *
+ * RESULT: Don't keep a duplicated structure around always. During block processing just push to the tree. During
+ * epoch processing some temporary flat structures are computed but dropped after processing the epoch.
+ */
+export type CachedBeaconState<T extends BeaconStateAllForks> = T & BeaconStateCache;
+
+export type CachedBeaconStatePhase0 = CachedBeaconState<BeaconStatePhase0>;
+export type CachedBeaconStateAltair = CachedBeaconState<BeaconStateAltair>;
+export type CachedBeaconStateBellatrix = CachedBeaconState<BeaconStateBellatrix>;
+export type CachedBeaconStateAllForks = CachedBeaconState<BeaconStateAllForks>;
+
+/**
+ * Create CachedBeaconState computing a new EpochContext instance
+ */
+export function createCachedBeaconState<T extends BeaconStateAllForks>(
+  state: T,
+  immutableData: EpochContextImmutableData,
+  opts?: EpochContextOpts
+): T & BeaconStateCache {
+  return getCachedBeaconState(state, {
+    config: immutableData.config,
+    epochCtx: EpochContext.createFromState(state, immutableData, opts),
+  });
+}
+
+/**
+ * Attach an already computed BeaconStateCache to a BeaconState object
+ */
+export function getCachedBeaconState<T extends BeaconStateAllForks>(
+  state: T,
+  cache: BeaconStateCache
+): T & BeaconStateCache {
+  const cachedState = state as T & BeaconStateCache;
+  cachedState.config = cache.config;
+  cachedState.epochCtx = cache.epochCtx;
+
+  // Overwrite .clone function to preserve cache
+  // TreeViewDU.clone() creates a new object that does not have the attached cache
+  const viewDUClone = cachedState.clone.bind(cachedState);
+
+  function clone(this: T & BeaconStateCache): T & BeaconStateCache {
+    const viewDUCloned = viewDUClone();
+    return getCachedBeaconState(viewDUCloned, {
+      config: this.config,
+      epochCtx: this.epochCtx.clone(),
+    }) as T & BeaconStateCache;
+  }
+
+  cachedState.clone = clone as typeof viewDUClone;
```

Comment:
This is a specially critical an awkward part of the state cache. It re-writes the clone method set by the TreeViewDU. Would be better to use [Object.defineProperty()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty)?